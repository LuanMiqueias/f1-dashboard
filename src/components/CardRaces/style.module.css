
.card_races{
  width: 100%;
  transition: all .2s;
  box-shadow: 2px 2px 5px 1px rgba(0, 0, 0, 0.15);
  width: 100%;
  display: flex;
  align-items: flex-end;
  gap:8px;
  justify-content: space-between;
  padding: 1rem 1rem;
  border-radius: 5px;
  cursor: pointer;
  background: var(--white);
  position: relative;
  user-select: none;
  border:1px solid rgb(238, 238, 238);
}
.card_races:hover{
  transition: all .07s;
  transform: translateX(-2%) ;
  box-shadow: 3px 3px 8px 1px rgba(0, 0, 0, 0.2);
}
.content_card_races{
  display: flex;
  align-items: flex-start;
  justify-content: flex-start;
  flex-direction: column;
  gap:4px
}
.content_card_races h2{
  font-weight: 400;
  font-size: 1rem;
  color:var(--sub-title);
  max-width: 120px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}
.content_card_races p{
  font-weight: 400;
  font-size: .75rem;
  color:var(--text-light);
  max-width: 220px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}