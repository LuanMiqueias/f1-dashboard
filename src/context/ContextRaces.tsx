import React, { ReactNode } from "react";
import api from "../services/api";

interface PropsRaces {
  id: number;
  distance?: string;
  date: string;
  status: string;
  season: number;
  type: string;
  laps: {
    total?: number;
  };
  circuit: {
    id: number;
    name: string;
    image: string;
  };
}
interface IData {
  data: PropsRaces[];
}
interface RacesContent {
  races: IData;
  idCompetition: number;
  oneRace: PropsRaces;
  loading: boolean;
  error: string;
  selectRace: (id: number) => void;
  updateRaces: (data: IData) => void;
  updateIdCompetition: (id: number) => void;
}
export const RacesContext = React.createContext({} as RacesContent);

export const RacesProvider: React.FC<ReactNode> = ({ children }) => {
  const [races, setRaces] = React.useState(null as IData);
  const [idCompetition, setIdCompetition] = React.useState(null as number);
  const [oneRace, setOneRace] = React.useState(null as PropsRaces);
  const [loading, setLoading] = React.useState(true as boolean);
  const [error, setError] = React.useState("" as string);

  React.useEffect(() => {
    //DeverÃ¡ set feito um Fetch para api usando o ID da competicao
    if (idCompetition) {
      setLoading(true);
      (async () => {
        const response = await api.get(
          `races?competition=${idCompetition}&season=2021`,
          {
            headers: {
              "x-rapidapi-host": "v1.formula-1.api-sports.io",
              "x-rapidapi-key": process.env.NEXT_PUBLIC_API_KEY,
            },
          }
        );
        if (response.data.errors.request) {
          setError(
            "Sorry for the inconvenience, this site uses a free API plan and you or someone else has exceeded the limit ðŸ˜­, please try again later."
          );
        }
        setRaces({ data: response.data.response });
        setLoading(false);
      })();
    }
  }, [idCompetition]);

  function updateRaces(data: IData) {
    setRaces(data);
    setOneRace(null);
  }
  function updateIdCompetition(id: number) {
    setIdCompetition(id);
    setOneRace(null);
  }
  function selectRace(id: number) {
    if (races) {
      const race = races.data.filter((item) => {
        return item.id === id;
      });
      setOneRace(race[0]);
    }
  }
  return (
    <RacesContext.Provider
      value={{
        races,
        idCompetition,
        oneRace,
        loading,
        error,
        selectRace,
        updateRaces,
        updateIdCompetition,
      }}
    >
      {children}
    </RacesContext.Provider>
  );
};
