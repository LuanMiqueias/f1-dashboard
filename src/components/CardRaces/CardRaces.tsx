import React from "react";
import { RacesContext } from "../../context/ContextRaces";
import RaceInfo from "../RaceInfo/RaceInfo";
import styles from "./style.module.css";

interface PropsRaces {
  id: number;
  date: string;
  type: string;
  circuit: {
    name: string;
  };
}
const CardRaces = ({ id, type, date, circuit }: PropsRaces) => {
  const { selectRace, oneRace } = React.useContext(RacesContext);
  function formatDate(date: string) {
    const arrayMonths = [
      "Jan",
      "Fev",
      "Mar",
      "Abr",
      "Mai",
      "Jun",
      "Jul",
      "Ago",
      "Set",
      "Out",
      "Nov",
      "Dev",
    ];
    const dateConvert = new Date(date);
    const month = arrayMonths[dateConvert.getMonth()];
    const day = dateConvert.getDate();
    const hours = dateConvert.getHours();
    const minutes = dateConvert.getMinutes();
    let type = "PM";
    if (hours > 0 && hours < 12) {
      type = "AM";
    }
    return `${month} ${day}, ${hours === 0 ? "00" : hours}:${
      minutes === 0 ? "00" : minutes
    } ${type}`;
  }
  return (
    <>
      <div
        className={styles.card_races}
        onClick={() => {
          selectRace(id);
        }}
      >
        <div className={styles.content_card_races}>
          <h2>{circuit.name}</h2>
          <p>{type}</p>
        </div>
        <div className={styles.content_card_races}>
          <p>{formatDate(date)}</p>
        </div>
      </div>
      {screen.width < 958 && <>{oneRace?.id === id && <RaceInfo />}</>}
    </>
  );
};

export default CardRaces;
