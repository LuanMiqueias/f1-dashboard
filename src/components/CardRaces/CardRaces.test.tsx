import { fireEvent, render } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { RacesContext, RacesProvider } from "../../context/ContextRaces";
import CardRaces from "./CardRaces";

const fakeData = {
  id: 1,
  date: "1995-12-17T03:24:00",
  type: "type test",
  circuit: {
    name: "name test",
  },
};
const racesFake = {
  data: [
    {
      id: 1,
      date: "1995-12-17T03:24:00",
      status: "string",
      season: 1,
      type: "string",
      laps: {
        total: 10,
      },
      circuit: {
        id: 1,
        name: "string",
        image: "string",
      },
    },
  ],
};
it("Should match snapshot", () => {
  const { container } = render(<CardRaces {...fakeData} />);

  expect(container.firstChild).toMatchSnapshot();
});

it("Should change the id on onclick", () => {
  const contextCallback = jest.fn();
  const { container } = render(
    <RacesProvider>
      <CardRaces {...fakeData} />
      <RacesContext.Consumer>{contextCallback}</RacesContext.Consumer>
    </RacesProvider>
  );
  act(() => {
    contextCallback.mock.calls[0][0].updateRaces(racesFake);
  });
  fireEvent.click(container.firstElementChild);
  expect(contextCallback.mock.calls[2][0].oneRace.id).toEqual(1);
});
