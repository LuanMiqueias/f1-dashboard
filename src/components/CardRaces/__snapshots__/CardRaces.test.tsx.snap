// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<div
  class="card_races"
>
  <div
    class="content_card_races"
  >
    <h2>
      name test
    </h2>
    <p>
      type test
    </p>
  </div>
  <div
    class="content_card_races"
  >
    <p>
      Dev 17, 3:24 AM
    </p>
  </div>
</div>
`;
